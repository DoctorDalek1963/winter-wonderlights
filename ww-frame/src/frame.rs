//! This module handles the implementations of the frames.

use crate::{FrameObject, RGBArray};
use serde::{Deserialize, Serialize};
use std::fmt;
use tracing::trace;
use ww_gift_coords::COORDS;

/// A type of frame data.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub enum FrameType {
    /// A frame indicating that all the lights are off.
    Off,

    /// The colour for each light on the chain.
    ///
    /// The driver is expected to handle the cases where the Vec is too long or too short and
    /// doesn't corespond one-to-one to the lights, but this behaviour is specific to individual
    /// drivers and should not be relied on.
    RawData(Vec<RGBArray>),

    /// A sophisticated 3D frame, made of several objects.
    Frame3D(Frame3D),
}

/// A 3D frame, made of several objects.
#[derive(Clone, PartialEq, Serialize, Deserialize)]
pub struct Frame3D {
    /// The vec of objects in the frame.
    ///
    /// An object later in the vec will override any object previous in the list. No blending of
    /// colours is performed.
    objects: Vec<FrameObject>,

    /// Whether to blend objects together.
    blend: bool,

    /// Optionally pre-computed data. If an effect needs to compute the raw data first, then it can
    /// just store that data here so the driver doesn't have to re-compute it later.
    #[cfg_attr(any(feature = "insta", test), serde(skip))]
    pre_computed_raw_data: Option<Vec<RGBArray>>,
}

impl fmt::Debug for Frame3D {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Frame3D")
            .field("objects", &self.objects)
            .field("blend", &self.blend)
            //.field("pre_computed_raw_data", &self.pre_computed_raw_data)
            .finish()
    }
}

impl Frame3D {
    /// Create a new frame.
    pub fn new(objects: Vec<FrameObject>, blend: bool) -> Self {
        Self {
            objects,
            blend,
            pre_computed_raw_data: None,
        }
    }

    /// Compute the vec of raw data for this frame, using [`struct@COORDS`] to know where the
    /// lights are. Use [`Self::to_raw_data`] to get the data out or [`Self::raw_data`] to
    /// reference the optional data.
    ///
    /// This function does nothing if the data was already computed.
    ///
    /// This function returns a reference to self to allow things like:
    /// ```
    /// #![feature(let_chains)]
    ///
    /// # use ww_frame::Frame3D;
    /// # fn do_something(_: &Vec<[u8; 3]>) -> bool { true }
    /// # let mut frame = Frame3D::new(vec![], false);
    /// while let Some(data) = frame.compute_raw_data().raw_data() && do_something(data) {
    ///     // Do stuff
    ///     # break;
    /// }
    /// ```
    /// Admittedly this API isn't very ergonomic and this function should ideally return a
    /// reference to the computed data, but the borrow checker didn't like that, so we're currently
    /// using this bodge.
    pub fn compute_raw_data(&mut self) -> &mut Self {
        if self.pre_computed_raw_data.is_some() {
            return self;
        }

        let mut data: Vec<RGBArray> = vec![[0, 0, 0]; COORDS.lights_num()];
        trace!(?data, "Before");

        if self.blend {
            let total = self.objects.len() as f32;

            data = self
                .objects
                .iter()
                // Render the data into its own slice and convert all the u8s to f32s for later
                .map(|f_obj| {
                    let mut new_data = data.clone();
                    f_obj.render_into_slice(&mut new_data);
                    new_data
                        .into_iter()
                        // Divide by the total here to make the mean easier to calculate
                        .map(|[r, g, b]| [r as f32 / total, g as f32 / total, b as f32 / total])
                        .collect::<Vec<_>>()
                })
                // Sum all the data to get one Vec<[f32; 3]> where each element is the mean colour
                // for that light
                .fold(vec![[0., 0., 0.]; COORDS.lights_num()], |mut acc, v| {
                    for i in 0..COORDS.lights_num() {
                        acc[i][0] += v[i][0];
                        acc[i][1] += v[i][1];
                        acc[i][2] += v[i][2];
                    }
                    acc
                })
                .into_iter()
                // Convert everything back to u8s
                .map(|[r, g, b]| [r as u8, g as u8, b as u8])
                .collect();
        } else {
            for frame_object in &self.objects {
                frame_object.render_into_slice(&mut data);
            }
        }

        trace!(?data, "After");
        self.pre_computed_raw_data = Some(data);
        self
    }

    /// Return the raw data for this frame.
    pub fn to_raw_data(mut self) -> Vec<RGBArray> {
        self.compute_raw_data();
        self.pre_computed_raw_data.expect(
            "pre_computed_raw_data must be populated since compute_raw_data() has been called",
        )
    }

    /// Get an optional reference to the internal pre-computed data.
    pub fn raw_data(&self) -> Option<&Vec<RGBArray>> {
        self.pre_computed_raw_data.as_ref()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Object;
    use glam::Vec3;

    #[test]
    fn to_raw_data_test() {
        let green_plane = FrameObject {
            object: Object::Plane {
                normal: Vec3::new(1., 0.5, -3.5).normalize(),
                k: -1.2354,
                threshold: 0.15,
            },
            colour: [25, 200, 16],
            fadeoff: 0.14,
        };
        let blue_plane = FrameObject {
            object: Object::Plane {
                normal: Vec3::new(-1.6, 1.99, -0.02).normalize(),
                k: 0.24,
                threshold: 0.3,
            },
            colour: [0, 12, 186],
            fadeoff: 0.82,
        };
        let red_sphere = FrameObject {
            object: Object::Sphere {
                center: Vec3::new(0.01, 0.31, 0.2),
                radius: 1.2,
            },
            colour: [243, 36, 17],
            fadeoff: 0.1,
        };

        // Single green plane

        let single_green_plane = Frame3D::new(vec![green_plane.clone()], false);
        insta::with_settings!({
            info => &single_green_plane,
            description => "Rendering a single plane with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(single_green_plane.to_raw_data());
        });

        // Green and blue planes

        let green_and_blue_planes_no_blend =
            Frame3D::new(vec![green_plane.clone(), blue_plane.clone()], false);
        insta::with_settings!({
            info => &green_and_blue_planes_no_blend,
            description => "Rendering two planes with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(green_and_blue_planes_no_blend.to_raw_data());
        });

        let green_and_blue_planes_with_blend =
            Frame3D::new(vec![green_plane.clone(), blue_plane.clone()], true);
        insta::with_settings!({
            info => &green_and_blue_planes_with_blend,
            description => "Rendering two planes with with blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(green_and_blue_planes_with_blend.to_raw_data());
        });

        // Blue plane and red sphere

        let blue_plane_red_sphere_no_blend =
            Frame3D::new(vec![blue_plane.clone(), red_sphere.clone()], false);
        insta::with_settings!({
            info => &blue_plane_red_sphere_no_blend,
            description => "Rendering a blue plane and then a red sphere with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(blue_plane_red_sphere_no_blend.to_raw_data());
        });

        let red_sphere_blue_plane_no_blend =
            Frame3D::new(vec![red_sphere.clone(), blue_plane.clone()], false);
        insta::with_settings!({
            info => &red_sphere_blue_plane_no_blend,
            description => "Rendering a red sphere and then a blue plane with no blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(red_sphere_blue_plane_no_blend.to_raw_data());
        });

        let blue_plane_red_sphere_with_blend =
            Frame3D::new(vec![blue_plane.clone(), red_sphere.clone()], false);
        insta::with_settings!({
            info => &blue_plane_red_sphere_with_blend,
            description => "Rendering a blue plane and then a red sphere with blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(blue_plane_red_sphere_with_blend.to_raw_data());
        });

        let red_sphere_blue_plane_with_blend =
            Frame3D::new(vec![red_sphere.clone(), blue_plane.clone()], false);
        insta::with_settings!({
            info => &red_sphere_blue_plane_with_blend,
            description => "Rendering a red sphere and then a blue plane with blend",
            omit_expression => true,
        }, {
            insta::assert_ron_snapshot!(red_sphere_blue_plane_with_blend.to_raw_data());
        });
    }
}
