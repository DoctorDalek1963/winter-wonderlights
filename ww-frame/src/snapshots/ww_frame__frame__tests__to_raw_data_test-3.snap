---
source: ww-frame/src/frame.rs
description: Rendering two planes with with blend
info:
  objects:
    - object:
        Plane:
          normal:
            - 0.27216554
            - 0.13608277
            - -0.9525794
          k: -1.2354
          threshold: 0.15
      colour:
        - 25
        - 200
        - 16
      fadeoff: 0.14
    - object:
        Plane:
          normal:
            - -0.6265843
            - 0.7793142
            - -0.007832304
          k: 0.24
          threshold: 0.3
      colour:
        - 0
        - 12
        - 186
      fadeoff: 0.82
  blend: true
---
[
  (0, 6, 93),
  (0, 5, 82),
  (0, 5, 88),
  (0, 5, 79),
  (0, 5, 89),
  (0, 5, 82),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 81),
  (0, 5, 91),
  (0, 5, 84),
  (0, 5, 82),
  (0, 5, 79),
  (0, 5, 88),
  (0, 5, 92),
  (0, 5, 92),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 92),
  (0, 5, 92),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 77),
  (0, 5, 82),
  (0, 4, 74),
  (0, 4, 70),
  (0, 3, 60),
  (0, 3, 50),
  (0, 3, 52),
  (0, 3, 60),
  (0, 2, 41),
  (0, 0, 11),
  (0, 0, 5),
  (0, 1, 16),
  (0, 1, 22),
  (0, 2, 32),
  (0, 1, 29),
  (0, 3, 53),
  (0, 4, 76),
  (0, 4, 71),
  (0, 4, 65),
  (0, 5, 85),
  (0, 6, 93),
  (0, 5, 81),
  (0, 4, 67),
  (0, 4, 68),
  (0, 4, 70),
  (0, 3, 47),
  (0, 3, 50),
  (0, 2, 31),
  (0, 2, 42),
  (0, 3, 49),
  (0, 2, 46),
  (0, 1, 19),
  (0, 0, 6),
  (0, 1, 29),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 12),
  (0, 1, 30),
  (0, 2, 43),
  (0, 3, 49),
  (0, 3, 56),
  (0, 4, 70),
  (0, 4, 75),
  (0, 6, 93),
  (0, 4, 70),
  (0, 3, 60),
  (0, 3, 55),
  (0, 2, 39),
  (3, 26, 33),
  (12, 102, 39),
  (0, 0, 1),
  (0, 0, 0),
  (0, 1, 25),
  (0, 0, 15),
  (0, 1, 27),
  (0, 2, 36),
  (0, 2, 37),
  (0, 3, 59),
  (0, 3, 55),
  (0, 3, 61),
  (0, 5, 84),
  (0, 3, 60),
  (0, 6, 63),
  (11, 99, 76),
  (7, 61, 55),
  (0, 4, 75),
  (0, 5, 79),
  (4, 44, 96),
  (11, 99, 100),
  (12, 106, 101),
  (2, 26, 94),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 88),
  (7, 66, 79),
  (0, 6, 93),
  (0, 5, 83),
  (7, 67, 77),
  (10, 90, 58),
  (0, 3, 54),
  (0, 2, 40),
  (0, 2, 37),
  (0, 3, 54),
  (0, 2, 38),
  (0, 3, 51),
  (0, 4, 72),
  (0, 4, 65),
  (0, 5, 79),
  (0, 3, 56),
  (0, 4, 69),
  (0, 4, 67),
  (0, 3, 46),
  (0, 4, 65),
  (0, 4, 74),
  (0, 4, 67),
  (0, 5, 80),
  (0, 3, 52),
  (0, 2, 44),
  (0, 3, 50),
  (0, 4, 75),
  (5, 49, 72),
  (0, 5, 88),
  (0, 6, 93),
  (1, 14, 93),
  (6, 60, 97),
  (0, 6, 93),
  (0, 6, 93),
  (1, 17, 93),
  (0, 6, 93),
  (0, 5, 88),
  (9, 85, 99),
  (12, 106, 101),
  (12, 106, 101),
  (12, 106, 101),
  (12, 106, 101),
  (12, 106, 101),
  (12, 102, 100),
  (1, 18, 93),
  (0, 5, 85),
  (5, 50, 83),
  (0, 3, 57),
  (0, 3, 48),
  (2, 26, 48),
  (6, 59, 67),
  (2, 18, 45),
  (0, 4, 66),
  (0, 3, 53),
  (0, 4, 70),
  (0, 3, 54),
  (0, 3, 55),
  (7, 61, 65),
  (11, 95, 77),
  (12, 104, 75),
  (12, 104, 75),
  (9, 78, 53),
  (0, 4, 68),
  (0, 5, 80),
  (2, 26, 94),
  (11, 94, 100),
  (12, 106, 101),
  (12, 106, 101),
  (0, 6, 93),
  (9, 78, 98),
  (0, 6, 93),
  (0, 4, 69),
  (0, 4, 75),
  (1, 17, 52),
  (12, 103, 66),
  (12, 103, 64),
  (9, 76, 80),
  (12, 104, 84),
  (12, 105, 89),
  (12, 106, 101),
  (12, 105, 96),
  (9, 82, 99),
  (0, 9, 93),
  (8, 76, 98),
  (7, 68, 98),
  (12, 106, 101),
  (12, 106, 101),
  (12, 106, 101),
  (2, 28, 94),
  (12, 106, 101),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 8, 93),
  (12, 106, 101),
  (12, 106, 101),
  (12, 106, 101),
  (12, 106, 101),
  (8, 71, 98),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 89),
  (0, 6, 93),
  (0, 6, 93),
  (2, 27, 80),
  (11, 97, 100),
  (12, 105, 94),
  (12, 106, 101),
  (12, 105, 93),
  (11, 97, 90),
  (12, 103, 69),
  (7, 66, 66),
  (12, 103, 65),
  (11, 98, 62),
  (12, 102, 53),
  (1, 15, 55),
  (12, 102, 53),
  (12, 102, 41),
  (12, 103, 62),
  (4, 44, 96),
  (12, 102, 54),
  (12, 102, 51),
  (10, 84, 52),
  (5, 49, 44),
  (0, 4, 69),
  (0, 3, 47),
  (0, 4, 75),
  (0, 3, 61),
  (0, 4, 72),
  (0, 3, 54),
  (0, 4, 70),
  (0, 4, 69),
  (0, 4, 66),
  (0, 4, 66),
  (0, 3, 58),
  (0, 4, 63),
  (0, 3, 58),
  (0, 3, 60),
  (0, 4, 63),
  (0, 3, 56),
  (0, 2, 38),
  (3, 32, 25),
  (7, 62, 22),
  (12, 101, 35),
  (12, 101, 36),
  (8, 72, 13),
  (9, 78, 31),
  (0, 1, 19),
  (0, 0, 14),
  (0, 0, 10),
  (0, 2, 31),
  (0, 0, 12),
  (0, 0, 11),
  (0, 0, 2),
  (0, 0, 0),
  (0, 0, 3),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 0),
  (0, 0, 11),
  (0, 0, 2),
  (0, 3, 46),
  (0, 1, 29),
  (0, 2, 31),
  (0, 2, 41),
  (0, 3, 46),
  (1, 16, 45),
  (0, 2, 43),
  (12, 102, 50),
  (12, 102, 48),
  (12, 102, 51),
  (12, 103, 58),
  (12, 101, 28),
  (12, 101, 23),
  (12, 102, 41),
  (0, 2, 34),
  (0, 3, 56),
  (0, 3, 51),
  (0, 2, 42),
  (0, 3, 50),
  (0, 2, 34),
  (0, 1, 16),
  (0, 1, 29),
  (12, 99, 39),
  (10, 87, 44),
  (12, 101, 29),
  (10, 89, 42),
  (0, 0, 11),
  (0, 1, 18),
  (0, 1, 21),
  (0, 0, 0),
  (0, 1, 21),
  (0, 1, 24),
  (0, 1, 30),
  (0, 1, 24),
  (0, 2, 34),
  (0, 2, 39),
  (0, 1, 29),
  (0, 3, 61),
  (0, 4, 63),
  (8, 69, 67),
  (12, 104, 74),
  (8, 70, 65),
  (0, 5, 80),
  (0, 3, 56),
  (0, 2, 38),
  (0, 4, 64),
  (4, 36, 66),
  (0, 2, 45),
  (0, 4, 67),
  (5, 46, 77),
  (4, 41, 75),
  (1, 17, 61),
  (12, 104, 73),
  (12, 104, 76),
  (12, 105, 96),
  (2, 23, 81),
  (2, 23, 91),
  (0, 6, 93),
  (0, 4, 75),
  (0, 6, 93),
  (0, 6, 93),
  (0, 4, 75),
  (0, 5, 86),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 91),
  (0, 3, 59),
  (0, 4, 65),
  (0, 4, 75),
  (0, 3, 56),
  (0, 3, 59),
  (1, 17, 47),
  (10, 90, 64),
  (9, 78, 65),
  (10, 89, 71),
  (12, 104, 78),
  (12, 105, 87),
  (12, 105, 88),
  (0, 3, 57),
  (0, 3, 59),
  (0, 3, 61),
  (0, 5, 81),
  (0, 5, 87),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 10, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 4, 77),
  (0, 4, 75),
  (0, 4, 67),
  (0, 5, 91),
  (0, 5, 90),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 79),
  (0, 5, 77),
  (0, 4, 63),
  (0, 4, 70),
  (0, 4, 73),
  (0, 4, 68),
  (0, 4, 68),
  (0, 4, 75),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 4, 74),
  (0, 5, 84),
  (0, 4, 63),
  (0, 4, 64),
  (0, 3, 52),
  (0, 2, 43),
  (0, 3, 50),
  (0, 2, 44),
  (0, 2, 36),
  (0, 2, 42),
  (0, 2, 37),
  (0, 4, 63),
  (0, 3, 48),
  (0, 4, 63),
  (0, 5, 81),
  (0, 4, 64),
  (0, 4, 70),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 88),
  (0, 5, 79),
  (0, 5, 81),
  (0, 4, 76),
  (0, 4, 67),
  (0, 4, 66),
  (0, 4, 64),
  (0, 4, 67),
  (0, 3, 52),
  (0, 5, 79),
  (0, 5, 86),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 84),
  (0, 4, 70),
  (0, 5, 90),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 6, 93),
  (0, 5, 86),
  (0, 6, 93),
  (0, 5, 83),
  (0, 4, 63),
  (0, 3, 56),
  (0, 2, 43),
  (0, 3, 60),
  (0, 2, 43),
  (0, 2, 45),
  (0, 2, 33),
  (0, 2, 46),
  (0, 2, 39),
]
